Sources/NetworkProtection/Networking/NetworkProtectionClient.swift

```diff
@@ -175,12 +191,16 @@ final class NetworkProtectionBackendClient: NetworkProtectionClient {
         return decoder
     }()
 
-    private let endpointURL: URL
+    private var endpointURL: URL {
+        settings.selectedEnvironment.endpointURL
+    }
+
     private let isSubscriptionEnabled: Bool
+    private let settings: VPNSettings
 
-    init(environment: VPNSettings.SelectedEnvironment = .default, isSubscriptionEnabled: Bool) {
+    init(settings: VPNSettings, isSubscriptionEnabled: Bool) {
```

Comment:
This change allowed us to instantiate the client so that we don’t need to re-init the `NetworkProtectionDeviceManager` every time it's used. Should make testing a bit easier further down the line + also means we can guarantee the client always gets the most up-to-date environment setting

Reply:
I also feel like replacing the `environment` param with `settings` should at least be considered outside the scope of this PR/branch. It already does a lot and refactoring this might lead to side effect.

Reply:
I didn’t fully explain the above. The reason why it’s important to not recreate the device manager every time is that we could end up with two register requests happening at the same time which could create racy behaviour. The `NetworkProtectionDeviceManager` is an actor and it makes sense to just have one instance which will queue up these requests rather than creating new instances every time start, stop or failure recovery happens.

Rreplacing the `environment` param with `settings` was so that we could still support changing the environment without needing to restart the PacketTunnelProvider, but @diegoreymendez and I had a chat and we think it’s actually probable more sensible to require restarting as the token we’re using would become invalid.

I’m going to revert the injection of `settings` instead of `environment`, but I will still create a single instance of the `NetworkProtectionDeviceManager`. This will mean that we need to manually restart the tunnel for the env change to take effect.